sfxclient/multitokensink.go

```diff
@@ -0,0 +1,734 @@
+func (a *asyncTokenStatusMap) Increment(status *tokenStatus) {
+	select {
+	case <-a.stop: // check if the counter has been stopped
+		return
+	default:
+		select {
+		case a.input <- status:
+		default:
+		}
+	}
+}
+
+func newAsyncTokenStatusMap(name string, buffer int, numWorkers int64, defaultDims map[string]string) *asyncTokenStatusMap {
+	a := &asyncTokenStatusMap{
+		name:              name,
+		dataStore:         map[string]map[string]int64{},
+		input:             make(chan *tokenStatus, int64(buffer)*numWorkers),
+		stop:              make(chan bool),
+		requestDatapoints: make(chan bool, 5000),
+		returnDatapoints:  make(chan []*datapoint.Datapoint, 5000),
+		defaultDims:       defaultDims,
+	}
+	go func() {
+		for {
+			select {
+			case <-a.stop: // signal for the goroutine to stop managing the datapoints object
+				a.fetchDatapoints() // fetch datapoints one last time and set it to previousVal so anyone reading can get it
+				close(a.returnDatapoints)
+				return
+			case input := <-a.input:
+				a.processInput(input)
+			outer:
+				for {
+					select {
+					case input := <-a.input:
+						a.processInput(input)
+					default:
+						break outer
+					}
+				}
+			case <-a.requestDatapoints:
+				a.returnDatapoints <- a.fetchDatapoints()
+			}
+		}
+	}()
+	return a
+}
+
+// worker manages a pipeline for emitting metrics
+type worker struct {
+	lock         *sync.Mutex       // lock to control concurrent access to the worker
+	errorHandler func(error) error // error handler for handling error emitting datapoints
+	sink         *HTTPSink         // sink is an HTTPSink for emitting datapoints to Signal Fx
+	closing      chan bool         // channel to signal that the worker is stopping
+	done         chan bool         // channel to signal that the worker is done
+}
+
+// returns a new instance of worker with an configured emission pipeline
+func newWorker(errorHandler func(error) error, closing chan bool, done chan bool) *worker {
+	w := &worker{
+		lock:         &sync.Mutex{},
+		sink:         NewHTTPSink(),
+		errorHandler: errorHandler,
+		closing:      closing,
+		done:         done,
+	}
+
+	return w
+}
+
+// worker for handling datapoints
+type datapointWorker struct {
+	*worker
+	input     chan *dpMsg // channel for inputing datapoints into a worker
+	buffer    []*datapoint.Datapoint
+	batchSize int
+	stats     *asyncMultiTokenSinkStats // stats about
+}
+
+// emits a series of datapoints
+func (w *datapointWorker) emit(token string) {
+	// set the token on the HTTPSink
+	w.sink.AuthToken = token
+	// emit datapoints and handle any errors
+	err := w.sink.AddDatapoints(context.Background(), w.buffer)
+	w.handleError(err, token, w.buffer)
+	// account for the emitted datapoints
+	w.stats.TotalDatapointsBuffered.Increment(int64(len(w.buffer) * -1))
+}
+
+func (w *datapointWorker) handleError(err error, token string, datapoints []*datapoint.Datapoint) {
+	status := &tokenStatus{
+		status: "unknown",
```

Comment:
why is it a string? and all the atoi? http/status.go has all the codes and texts.  feels like givng unknown -1 or something and keeping the status code would allow you to use the statusText (in http/status.go) map better?